### PR TITLE
Update PoH speed check to derive rate from Bank

### PR DESCRIPTION
**Pull request by steviez**
#### Problem
The PoH speed check currently determines the target hash rate by reading values from genesis. However, the hashes per tick rate has been increased via feature gates. Thus, the speed check is using a slower hash rate for comparison than what is actually active.

#### Summary of Changes
So, update the PoH speed check to derive PoH hash rate from a Bank instead.

Running this on MNB, I see the following log emitted:
```
[2024-08-05T18:09:44.850926545Z INFO  solana_core::validator] PoH speed check: computed hashes per second 13917067, target hashes per second 10000000
```
10 million is the correct value; the log output shared in my previous PR (https://github.com/anza-xyz/agave/pull/2400) will show that we were previously showing 2 million as the target.

### CODE REVIEW

File: core/src/validator.rs [hunk 1]
```diff
@@ -1677,18 +1681,16 @@ fn active_vote_account_exists_in_bank(bank: &Bank, vote_account: &Pubkey) -> boo
...
+    if let Some(hashes_per_tick) = bank.hashes_per_tick() {
+        let ticks_per_slot = bank.ticks_per_slot();
         let hashes_per_slot = hashes_per_tick * ticks_per_slot;
         let hash_samples = maybe_hash_samples.unwrap_or(hashes_per_slot);
 
         let hash_time = compute_hash_time(hash_samples);
         let my_hashes_per_second = (hash_samples as f64 / hash_time.as_secs_f64()) as u64;
-        let target_slot_duration = Duration::from_nanos(genesis_config.ns_per_slot() as u64);
+
+        let target_slot_duration = Duration::from_nanos(bank.ns_per_slot as u64);
```
Comment (steviez) [inline core/src/validator.rs]: Not really sure `ns_per_slot` should be public in `Bank`, but seemingly a refactor for after this:
https://github.com/anza-xyz/agave/blob/4e5af1e90b0efee2d06c6b6ea63adc063cda2b10/runtime/src/bank.rs#L730

File: core/src/validator.rs [hunk 2]
```diff
@@ -2937,11 +2938,25 @@ mod tests {
...
+        assert_eq!(target_tick_duration_us, 6250);
+        Duration::from_micros(target_tick_duration_us)
+    }
+
     #[test]
     fn test_poh_speed() {
         solana_logger::setup();
         let poh_config = PohConfig {
-            target_tick_duration: Duration::from_millis(solana_sdk::clock::MS_PER_TICK),
+            target_tick_duration: target_tick_duration(),
```
Comment (steviez) [inline core/src/validator.rs:2960]: Technically, this change isn't critical for the PR. But, the constant `MS_PER_TICK` is 6ms; 6 ms/tick * 64 ticks/slot = 384 ms/slot. It annoyed me that this didn't come out to the 400ms, and I noticed when observing the `info` log from speed check. Can revert it back if desired

File: core/src/validator.rs [hunk 3]
```diff
@@ -1677,18 +1681,16 @@ fn active_vote_account_exists_in_bank(bank: &Bank, vote_account: &Pubkey) -> boo
...
 }
 
-fn check_poh_speed(
-    genesis_config: &GenesisConfig,
-    maybe_hash_samples: Option<u64>,
-) -> Result<(), ValidatorError> {
-    if let Some(hashes_per_tick) = genesis_config.hashes_per_tick() {
-        let ticks_per_slot = genesis_config.ticks_per_slot();
+fn check_poh_speed(bank: &Bank, maybe_hash_samples: Option<u64>) -> Result<(), ValidatorError> {
+    if let Some(hashes_per_tick) = bank.hashes_per_tick() {
```
Comment (bw-solana) [inline core/src/validator.rs]: Could avoid an extra indentation by possibly early returning here
```rust
let Some(hashes_per_tick) = bank.hashes_per_tick() else {
    warn!("");
    return Ok(());
}
...
```

**Review by bw-solana — DISMISSED** [review]
LGTM. Left one suggestion that would reduce indent level

**Review by steviez — COMMENTED** [review]
Ended up doing the let-else change you mentioned as I also thought this change might be worthy of a CHANGELOG entry (since it could now be an error at startup for nodes that can't hash fast enough)

**Review by bw-solana — APPROVED** [review]
LGTM

**Pull request merged**